source/connection_manager.c

```diff
@@ -0,0 +1,876 @@
+    AWS_ZERO_STRUCT(connections_to_release);
+
+    struct aws_linked_list pending_acquisitions_to_fail;
+    aws_linked_list_init(&pending_acquisitions_to_fail);
+
+    bool should_destroy = false;
+
+    AWS_LOGF_INFO(AWS_LS_HTTP_CONNECTION_MANAGER, "id=%p: release", (void *)manager);
+
+    aws_mutex_lock(&manager->lock);
+
+    if (manager->external_ref_count > 0) {
+        manager->external_ref_count -= 1;
+
+        if (manager->external_ref_count == 0) {
+            AWS_LOGF_INFO(
+                AWS_LS_HTTP_CONNECTION_MANAGER,
+                "id=%p: ref count now zero, starting shut down process",
+                (void *)manager);
+            manager->state = AWS_HCMST_SHUTTING_DOWN;
+            should_destroy = s_aws_http_connection_manager_should_destroy(manager);
+
+            /*
+             * swap our internal connection set with the zeroed local set
+             */
+            aws_array_list_init_dynamic(
+                &connections_to_release, manager->allocator, 0, sizeof(struct aws_http_connection *));
+            aws_array_list_swap_contents(&manager->connections, &connections_to_release);
+
+            /*
+             * Swap our pending acquisitions with the local list
+             */
+            aws_linked_list_swap_contents(&manager->pending_acquisitions, &pending_acquisitions_to_fail);
+
+            AWS_LOGF_INFO(
+                AWS_LS_HTTP_CONNECTION_MANAGER,
+                "id=%p: manager release, failing %zu pending acquisitions",
+                (void *)manager,
+                manager->pending_acquisition_count);
+            manager->pending_acquisition_count = 0;
+        }
+    } else {
+        AWS_LOGF_ERROR(
+            AWS_LS_HTTP_CONNECTION_MANAGER,
+            "id=%p: Connection manager release called with a zero reference count",
+            (void *)manager);
+    }
+
+    aws_mutex_unlock(&manager->lock);
+
+    size_t connection_count = aws_array_list_length(&connections_to_release);
+    if (connection_count > 0) {
+        AWS_LOGF_INFO(
+            AWS_LS_HTTP_CONNECTION_MANAGER,
+            "id=%p: manager release, releasing %zu held connections",
+            (void *)manager,
+            connection_count);
+    }
+
+    for (size_t i = 0; i < connection_count; ++i) {
+        struct aws_http_connection *connection = NULL;
+        if (aws_array_list_get_at(&connections_to_release, &connection, i)) {
+            continue;
+        }
+
+        manager->function_table->release_connection(connection);
+    }
+
+    aws_array_list_clean_up(&connections_to_release);
+    s_aws_http_connection_manager_complete_acquisitions(&pending_acquisitions_to_fail, manager->allocator);
+
+    if (should_destroy) {
+        s_aws_http_connection_manager_destroy(manager);
+    }
+}
+
+static void s_aws_http_connection_manager_build_task_set(
+    struct aws_http_connection_manager *manager,
+    struct aws_linked_list *completions,
+    size_t *new_connections) {
```

Comment:
should `new_connections` get zeroed out? Its value is only manipulated from one branch

Reply:
We always zero it before invoking this function but that wouldn't stop us from accidentally forgetting if things changed in the future, so I'll do it there as well.